### ISSUE TITLE
Output as NetCDF3 and variable sizes violate format constraints

### ISSUE THREAD

**Issue by bennyistanto** [monocongo/climate_indices#547]
Hi James @monocongo, I need help and advise,

I am using the latest package, installed using `pip install climate-indices` on a Conda env with python 3.11 inside WSL2 with Debian 12.

I try to calculate global SPI using TerraClimate data (4km spatial resolution - lon/lat 8640 4320), which break into 8 equal tiles (lon/lat 4320 1080). The file sizes is vary from the smallest around 500MB into the largest 3GB.

Largest file size
![Screenshot 2024-04-28 225842](https://github.com/monocongo/climate_indices/assets/722584/273c371e-fdb5-4e29-8687-aaf6ee0a7db7)

Smallest file size
![Screenshot 2024-04-28 225855](https://github.com/monocongo/climate_indices/assets/722584/0aa36e12-315a-4294-8b60-0941dff3e447)

Everything is working well, I completed all the SPI calculation for each tile using below, and produce an output with file size 28GB per file per scale per tile.
```bash
spi --periodicity monthly --scales 1 2 3 6 9 12 18 24 36 48 60 72 --calibration_start_year 1991 --calibration_end_year 2020 --netcdf_precip ppt00006.nc --var_name_precip ppt --output_file_base ../spi/tile06 --multiprocessing all
```

I checked the output using Panoply also looks good.
![Screenshot 2024-04-28 231812](https://github.com/monocongo/climate_indices/assets/722584/dd37083b-4581-4b12-989b-549483f6a85d)

Below is the metadata

```
File "tile03_spi_gamma_3_month.nc"
File type: NetCDF-3/CDM
 
netcdf file:/D:/temp/terraclimate/spi/tile03_spi_gamma_3_month.nc {
  dimensions:
    lat = 1080;
    lon = 4320;
    time = 792;
  variables:
    double lon(lon=4320);
      :standard_name = "longitude";
      :long_name = "longitude";
      :units = "degrees_east";
      :axis = "X";
      :_FillValue = NaN; // double

    double spi_gamma_3_month(lat=1080, lon=4320, time=792);
      :long_name = "Standardized Precipitation Index (Gamma), 3-month";
      :valid_min = -3.09; // double
      :valid_max = 3.09; // double
      :_FillValue = NaN; // double

    double lat(lat=1080);
      :standard_name = "latitude";
      :long_name = "latitude";
      :units = "degrees_north";
      :axis = "Y";
      :_FillValue = NaN; // double

    double time(time=792);
      :standard_name = "time";
      :long_name = "time";
      :axis = "T";
      :units = "days since 1900-01-01";
      :calendar = "gregorian";
      :_FillValue = NaN; // double

  // global attributes:
  :description = "SPI for 3-month scale computed from monthly precipitation input by the climate_indices package available from https://github.com/monocongo/climate_indices.";
  :geospatial_lat_min = 0.020833333333342807; // double
  :geospatial_lat_max = 44.97916666666667; // double
  :geospatial_lat_units = "degrees_north";
  :geospatial_lon_min = 0.020833333333342807; // double
  :geospatial_lon_max = 179.97916666666666; // double
  :geospatial_lon_units = "degrees_east";
}
```

The problem arise when I try to combine all tiles together to get the global coverage by: re-order the dimension from `lat,lon,time` to `time,lat,lon`, then follow with `collgrid` command using CDO. I just realized the output is NetCDF-3 which has maximum file size 2GB.

The error `**OSError: [Errno -62] NetCDF: One or more variable sizes violate format constraints**` continues to suggest a problem with handling the data within the NetCDF-3 format due to its inherent limitations. It seems the script is failing right at the attempt to open the NetCDF-3 file for conversion, indicating that the file might already be exceeding the constraints that NetCDF-3 can handle.

I tried with many approach, all is failed with the same error. 

* I can't check using `ncdump -h in.nc`

    ```bash
    ❯ ncdump -h /mnt/d/temp/terraclimate/spi/tile00_spi_gamma_3_month.nc
    ncdump: /mnt/d/temp/terraclimate/spi/tile00_spi_gamma_3_month.nc: NetCDF: One or more variable sizes violate format constraints
    ```

* Convert using nccopy also failed
 
    ```bash
    ❯ nccopy -k 4 tile00_spi_gamma_3_month.nc ../temp/tile00_spi_gamma_3_month.nc4
    NetCDF: One or more variable sizes violate format constraints
    Location: file ?; fcn ? line 2016
    ```

* I try to reorder the dimension directly using `ncpdq`, aslo failed with same reason

    ```bash
    ❯ for fl in *.nc; do ncpdq -a time,lat,lon $fl ../nc_tll/$fl; done
    ERROR: nco__open() unable to open file "tile00_spi_gamma_3_month.nc"
    ERROR NC_EVARSIZE One or more variable sizes violate format constraints
    ```

* I use `xarray` to convert from NetCDF-3 to NetCDF-4, then reorder the dimension, also got the same error

    ```python
    import os
    import xarray as xr
    import netCDF4 as nc
    import shutil
    
    # Define the directory containing the original NetCDF files
    input_directory = '/mnt/d/temp/terraclimate/spi'
    # Define the directory to save the reordered NetCDF files
    output_directory = '/mnt/d/temp/terraclimate/spi_tll'
    # Temporary directory for conversion
    temp_directory = '/mnt/d/temp/terraclimate/temp'
    
    # Make sure the output and temporary directories exist, create if they don't
    os.makedirs(output_directory, exist_ok=True)
    os.makedirs(temp_directory, exist_ok=True)
    
    def convert_to_netcdf4(input_path, output_path):
        """Converts an NetCDF-3 file to NetCDF-4 format."""
        with nc.Dataset(input_path, 'r') as src:
            with nc.Dataset(output_path, 'w', format='NETCDF4') as dst:
                # Copy dimensions
                for name, dimension in src.dimensions.items():
                    dst.createDimension(name, (len(dimension) if not dimension.isunlimited() else None))
                # Copy variables
                for name, variable in src.variables.items():
                    x = dst.createVariable(name, variable.datatype, variable.dimensions)
                    dst[name][:] = src[name][:]
                    # Copy variable attributes
                    dst[name].setncatts(src[name].__dict__)
                # Copy global attributes
                dst.setncatts(src.__dict__)
    
    for filename in os.listdir(input_directory):
        if filename.endswith('.nc'):
            input_file_path = os.path.join(input_directory, filename)
            temp_file_path = os.path.join(temp_directory, filename.replace('.nc', '.nc4'))
            output_file_path = os.path.join(output_directory, filename)
    
            # Convert to NetCDF-4 and store temporarily
            convert_to_netcdf4(input_file_path, temp_file_path)
    
            # Open the temporary NetCDF-4 file
            ds = xr.open_dataset(temp_file_path, chunks={'time': 10, 'lat': 432, 'lon': 432})
    
            # Reorder the dimensions if 'spi_gamma_3_month' variable exists
            if 'spi_gamma_3_month' in ds.variables:
                ds_reorder = ds.transpose('time', 'lat', 'lon')
                ds_reorder['spi_gamma_3_month'].attrs['dimensions'] = 'time lat lon'
                # Save the reordered dataset to a new NetCDF file
                ds_reorder.to_netcdf(output_file_path, format='NETCDF4', engine='netcdf4')
                print(f"Reordering complete and file saved to {output_file_path}")
    
            # Close the dataset and remove the temporary file
            ds.close()
            os.remove(temp_file_path)
    
    # Optionally, remove the entire temporary directory if empty
    shutil.rmtree(temp_directory, ignore_errors=True)
    
    print("Processing of all files complete.")
    ```

    And below is the error

    ```
    ---------------------------------------------------------------------------
    OSError                                   Traceback (most recent call last)
    Cell In[3], line 40
         37 output_file_path = os.path.join(output_directory, filename)
         39 # Convert to NetCDF-4 and store temporarily
    ---> 40 convert_to_netcdf4(input_file_path, temp_file_path)
         42 # Open the temporary NetCDF-4 file
         43 ds = xr.open_dataset(temp_file_path, chunks={'time': 10, 'lat': 432, 'lon': 432})
    
    Cell In[3], line 19, in convert_to_netcdf4(input_path, output_path)
         17 def convert_to_netcdf4(input_path, output_path):
         18     """Converts an NetCDF-3 file to NetCDF-4 format."""
    ---> 19     with nc.Dataset(input_path, 'r') as src:
         20         with nc.Dataset(output_path, 'w', format='NETCDF4') as dst:
         21             # Copy dimensions
         22             for name, dimension in src.dimensions.items():
    
    File src/netCDF4/_netCDF4.pyx:2469, in netCDF4._netCDF4.Dataset.__init__()
    
    File src/netCDF4/_netCDF4.pyx:2028, in netCDF4._netCDF4._ensure_nc_success()
    
    OSError: [Errno -62] NetCDF: One or more variable sizes violate format constraints: '/mnt/d/temp/terraclimate/spi/tile00_spi_gamma_3_month.nc'
    ```

My question is, any suggestion what should I do to solve this problem? is there any other approach to easily convert from v3 to v4?

**Comment by monocongo** [comment]
Is the issue that our package creates files that exceed the 2G size constraint of NetCDF? If so, this is a limitation (bug?) of the NetCDF (or xarray) package, in that it allows this to happen in the first place. We could add some code to our package to check that the provided input(s) won't result in output files that exceed the 2G limit.

**Comment by monocongo** [comment]
Is it possible to resize the inputs into smaller tiles, so they'll always fall below the constraint?

**Comment by bennyistanto** [comment]
I think I (just) found the problem this afternoon. Following the [guideline](https://climate-indices.readthedocs.io/en/latest/), after creating the new env, we install the `climate-indices` via pip then follow with installing NCO using Conda.

Using this approach, all the output will force to save as NetCDF-3 (I guess and I don't know if `xarray` will save into NetCDF-4 using standard `dataset.to_netcdf` without mentioning the format and engine `engine='netcdf4', format='NETCDF4',`), and the writing process for the output always succeed (in my case 28GB, beyond the 2GB). The problem arise when we try to access the output, it will give `Errno -62`.

Then I try to install `netcdf4` using Conda, and run the SPI calculation again, everything looks good, the SPI output is in NetCDF-4 (although the file size still 28GB). After re chunking using `ncks` and re order the dimension to `time,lat,lon` using `ncpdq`, the final tile output size reduce into 6GB only for areas lon/lat 4320 1080.

```
for fl in *.nc4; do ncks --cnk_dmn time,10 --cnk_dmn lat,180 --cnk_dmn lon,360 -4 "$fl" temp_"$fl" && ncpdq -a time,lat,lon temp_"$fl" ../spi_tll/"$fl" && rm temp_"$fl"; done
```

I tested above only for 1 tile and seems the output as expected. I will run for the rest (the completed output in NetCDF-3 takes 2 days). Well it such a wasting time, resources and storage due to forgot to install netcdf4.

Closing the issue.
